doc/adr/0007-introduce-doctrees-to-describe-non-api-docs.md

```diff
@@ -0,0 +1,158 @@
+such README should be taken into consideration as well.
+
+For a first version Markdown-only would be an acceptable limitation
+but extending to support other markup languages such as Asciidoc
+should be possible.
+
+#### Grimoire
+
+Since we use Grimoire as a storage system it would be nice to
+integrate this non-API information into the Grimoire store as well.
+A few challenges arise:
+
+- Grimoire's support for articles and similar kinds of documentation is unfinished
+- How would versioning work? This becomes especially tricky if we want
+  to update non-API documentation as soon as it is changed in Git.
+- How could a hierarchy of documents be encoded in Grimoire? Or should it even?
+
+
+## Decision
+
+We introduce a way for library authors to specify a hierarchy of
+various documentation files they store in their Git repository. This
+hierarchy can be encoded in familiar, hiccup-style vectors. A
+simplified example, assuming some Markdown files in `doc/`:
+
+```clojure
+[["Getting Started" {:file "doc/getting-started.md"}]
+ ["Guides" {}
+  ["Integrating Authentication" {:file "doc/integrating-auth.md"}]
+  ["Websockets" {:file "doc/websockets.md"}]]]
+```
+
+This information can then be used by cljdoc to create a manual or
+book-like experience. This approach is partly inspired by [Sphinx' TOC trees](http://www.sphinx-doc.org/en/stable/markup/toctree.html).
+
+Because Grimoire's support for articles is still in development
+this ADR does not contain a decision on how to store the information
+that cljdoc pulls out of projects using DocTrees.
+
+#### More Details on DocTrees
+
+A data structure may describe documentation hierarchy in the following way:
+
+- ordered vector of vectors, somewhat hiccup inspired
+- an attribute-map describes how the content of the given page can be
+  retrieved from the Git repository
+- elements after attribute-map are interpreted as children
+
+This data structure may be specified by library authors through
+a configuration file or derived from the repository structure.
+
+```clojure
+;; README only, perhaps derived
+[["Readme" {:file "README.md"}]]
+```
+
+```clojure
+;; Simple markdown files in doc/ — same as above
+[["Getting Started" {:file "doc/getting-started.md"}]
+ ["Guides" {}
+  ["Integrating Authentication" {:file "doc/integrating-auth.md"}]
+  ["Websockets" {:file "doc/websockets.md"}]]]
+```
+
+```clojure
+;; Sections in a single asciidoc file
+[["Getting Started" {:file "doc/content.adoc#getting-started"}]
+ ["Guides" {}
+  ["Integrating Authentication" {:file "doc/content.adoc#integrating-auth"}]
+  ["Websockets" {:file "doc/content.adoc#websockets"}]]]
+```
+
+## Consequences
+
+#### Monolithic Documentation
+
+Projects like
+[Fulcro](https://github.com/fulcrologic/fulcro/blob/develop/DevelopersGuide.adoc)
+have one big file which contains all documentation. It would therefore
+not be possible for Fulcro to describe a DocTree with items pointing
```

Comment:
Could Fulcro describe the doctree with hash links to section ids? That's a little bit web centric, but could be an option.

Reply:
I considered that and it would be possible. It would would add quite a bit of complexity though as we'd need to "chunk" documents, and therefore have a deeper understanding of the various document types we want to support (Markdown, Asciidoc, ...). Also in contrast to files uniqueness is not guaranteed.

Definitely willing to do this kind of thing if that's whats needed but for v1 I'll see what people say about splitting their docs into multiple files 🙂